### DISCUSSION TITLE
[Feature] Full-resolution/original image viewing setting or mode

### DISCUSSION THREAD

**Discussion by MaxLyt** [immich-app/immich#6306] [Feature Request]
### The feature

It would be great to allow viewing compatible photos (and videos) in full resolution rather than jpeg thumbnails. It's already possible to accomplish via zoom in and then zoom out - the image shown then is not the lower-res thumbnail but full-resolution original. I am suggesting a setting that would load originals by default. 

The simplest implementation would be a system-level toggle, but user-level or even client-device-level (e.g., thumbnails on phone,  full-res on desktop) settings would be progressively more useful. For videos, the system-level effect can already be accomplished by disabling transcoding in settings. 

System or user level settings should be rather simple to add, and it would add a lot of value to scenarios where a user wants to see/show content in its full quality. It's particularly important for HDR photos that get distorted by conversion to jpeg thumbnails (also, the tone mapping of that conversion is currently done rather poorly).

Further or alternative potential improvements could be:
1. An option to disable thumbnail creation for commonly supported file formats - similar to how it's already possible to disable transcoding for video.
2. Generating thumbnails on demand, based on the viewport size and format support of the client
3. Supporting a more current format for thumbnails, like AVIF - jpeg lacks support for higher bit depth or HDR

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by mertalev** [comment]
1. It's possible to allow more customization here, but the other part of this is that the images Immich generates have EXIF metadata stripped. We'd need to make it clear there are privacy / security implications to always showing the original. For instance, we don't allow the original image to be served if it's shared unless downloading is enabled for it.
2. The latency this would add would make for a poor user experience
3. I'd like to add AVIF, but it depends on the mobile side of things. Flutter still doesn't officially support AVIF, so we'd have to use a third-party package like flutter_avif for this format.

**Reply by MaxLyt** [reply]
All makes sense, thanks! My use case is not for sharing but for users viewing/showing on a big screen, so I did not think through sharing implications. For own viewing, when zooming in it still loads the original with EXIF intact, as far as I can tell. It even loads my original AVIFs in all their HDR glory, but it's a pain to zoom in and out for every one of them.

**Reply by NicholasFlamy** [reply]
I've heard that WEBP is pretty good. And with Google as the creator it should be supported on Android and likely iOS.

**Reply by MaxLyt** [reply]
WebP or not is somewhat beside the point - if there is a setting to load original images instead of thumbnails, you just get whatever images you save.

I'll probably have to do it myself once I find time because JPEG thumbnails make it completely unusable for my use case.

**Reply by NicholasFlamy** [reply]
The mobile app allows loading the originals.

**Reply by NicholasFlamy** [reply]
![Screenshot_20240129_195501](https://github.com/immich-app/immich/assets/30300649/1c6165c1-b16e-454e-9829-fe114e9d3cba)

**Reply by MaxLyt** [reply]
Yep. Now if there was a similar setting in the Web app, it would have been great. I don't understand the logic of making it a setting in the mobile app but choosing to make it on-zoom-only action for the Web app. I found a discussion of making this a setting for the Web app, and the "solution" was to load originals on zoom. But zooming on every image just to see it instead of thumbnail is very inconvenient.

**Comment by NicholasFlamy** [comment]
> 1. disable thumbnail creation

~~Set this to 0
![Screenshot_20240129_195920_Chrome](https://github.com/immich-app/immich/assets/30300649/6a05a7be-900d-4042-bca6-c920c922aeb6)~~

**Reply by MaxLyt** [reply]
Thanks, I hope this works! I'll try later today. I'm not sure if not having thumbnails will lead to the app using originals for viewing. Also, the small thumbs (the Webp ones) are still needed for the image grid.

**Reply by blasterspike** [reply]
If you try to set that to 0 you get the error
```
job.thumbnailGeneration.concurrency must be a positive number (Immich Server Error)
```

**Reply by NicholasFlamy** [reply]
Oh crap, sorry about that, that doesn't work.

**Comment by blasterspike** [comment]
I would love the option to see the originals instead of the large thumbnail!
This would be beneficial even for the directory size of Immich.

I have just installed Immich and I have giving it a try with an external library. I have disabled video transcoding, so the only files that are in _encoded-video/_ should be iPhone live photos converted if I'm not mistaken.
With a library of 71.04GB and 10.980 files I have a directory size for Immich (_encoded-video/_, _library/_, _thumbs/_) of 10.82GB 66.989 files. I'm expecting so many files because of the thumbnails but I would like to see the total size reduced.

**Comment by genaumann** [comment]
I am currently unable to view my HDR photos as the preview image is displayed in SDR.
After zooming in, the HDR takes effect.
It would be much nicer if I didn't have to zoom in to see the photo in HDR.

**Reply by mertalev** [reply]
There's an option to prefer the original in the user settings.

**Reply by genaumann** [reply]
Thanks, didn't know that.

**Comment by 1298345095** [comment]
Can you add a button to show the original photo? Now I must download it and look at it in the system photo app. It's disturbing...

**Reply by bo0tzz** [reply]
This button in the web viewer loads the original 
![image](https://github.com/user-attachments/assets/b6689516-0c96-4e56-baee-8cdd62d94c69)

**Reply by 1298345095** [reply]
> This button in the web viewer loads the original
> ![image](https://private-user-images.githubusercontent.com/11602424/351704780-b6689516-0c96-4e56-baee-8cdd62d94c69.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE4MjMwNTcsIm5iZiI6MTcyMTgyMjc1NywicGF0aCI6Ii8xMTYwMjQyNC8zNTE3MDQ3ODAtYjY2ODk1MTYtMGM5Ni00ZTU2LWJhZWUtOGNkZDYyZDk0YzY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI0VDEyMDU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlYzJjYzk0ODAyY2VkZDQ1YWI0NTI0MjkwY2JiNWY4NTJiNDYyYTM0YmMwYjY2M2NjNWNkMGVhMjYyYWU3ZDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QGpwQQwM3mcHB9dhVWfAs0ykvRQg6LnBKd-Zvz8UPnI)

I mean in ios app

**Reply by ovizii** [reply]
> This button in the web viewer loads the original !

I can't believe I just googled and experimented for 2h until I found your post. This is the most unintuitive icon ever. Who would have guessed, that clicking the zoom button, zooms in one level and loads the original in the background. I would have never noticed had I not read your Post here. Also, if you right-click a preview image and select “open in new tab” you'll see the preview image but if you click this zoom button first, you'll get the original in the new tab.

**Comment by antonio-ivanovski** [comment]
It would be super usefull to be able to load images/video adhoc for only one time. I want to see my 4k 60fps video sometimes in their full glory and not in the 1080p transcoded quality, without having to wait for the full video download.

This is useful because the video can be streamed and it is not downloaded fully on device taking up memory.

![image](https://github.com/user-attachments/assets/a9a1b7da-d4d7-41a3-8199-278806e29812)

**Reply by dJani97** [reply]
I would also love a full-res slideshow feature as well. I just started using Immich, and based on my testing so far, slideshows show low-res assets, unless the user zooms in. This is not ideal for presenting photos on a TV, for example.

**Reply by NicholasFlamy** [reply]
Personally, I think the slideshow feature should by default load original assets regardless of other settings, I'm surprised it doesn't.

**Comment by jhbruhn** [comment]
Related to this but not directly: When I upload videos in 4K60fps H264, they are not transcoded to my set-up 720p, but the original is always shown, which leads to very long load times (if connected remotely). The transcoding settings do not allow to set the target codec to also need transcoding. It would be nice if I could still enable transcoding for this case, to downsample the preview videos (and download/preview the original as mentioned above).

**Reply by NicholasFlamy** [reply]
I think you can accomplish that by doing this:
![Screenshot_20240924_111401_Brave](https://github.com/user-attachments/assets/e881a4ce-95b7-4163-b060-1207de794887)

**Reply by jhbruhn** [reply]
True, thanks!

**Comment by Darkrael** [comment]
I've had the same idea for some time now and since it does not look like this feature is getting implemented soon, i've done my best and scripted it in browser and it works quite well for my use case.  Posting it here if anyone wants a "hacky" solution for the browser in the interim: https://greasyfork.org/en/scripts/519391-immich-quality-switcher

**Reply by NicholasFlamy** [reply]
This seems interesting, will note that based on simply looking over it this has the same effect of zooming in and back out (I have not looked at the implementation though so I don't know how this script accomplishes this).

Note for anyone who see this:
If someone wants to see the original image by default this is now available in user settings as mentioned in my comment below.

**Reply by Darkrael** [reply]
Yes, for images technically does the same, however it saves the current state (original quality or preview quality) until reloads and applies it when navigating to other images. The state will also be kept consistent for zooming in and out (so by default zooming in will keep the preview quality). 
The big difference is that i also implemented it for videos. This is the important part for me since i have them compressed quite a bit for viewing on the go due to low upload speed on my connection and i wanted a way to quickly view the original when im at home and have a direct connection to the server.
For a quick technical explanation: The script replaces the `src` parameter on images from the substring `thumbnail?size=preview&` to `original?` and on video from `video/playback` to `original` which will point to the original versions of the files. It is also using MutationObserver to find new nodes (since immich replaces content when navigating instead of loading a new URL) and apply the correct transformation depending on the current state.
I also added a note on greasyfork mentioning that you can toggle the default behavior by changing the global variable `SHOW_ORIGINAL` in line 18 to `true`

**Reply by NicholasFlamy** [reply]
This is very clever. I like it. (Note: haven't tried it yet but it's cool.)

This might be the kind of thing you could add to the community projects. You could add a link to your project here: https://immich.app/docs/community-projects

(If you do decide to PR this page with your project, I'll happily try the project and confirm that it works as expected and help you get your PR merged.)


Edit: Thanks for the technical explanation too. Forgot to thank you for that specifically, I like knowing how things work!

**Comment by NicholasFlamy** [comment]
> I am suggesting a setting that would load originals by default.

This was available on mobile at an earlier time which can be seen in an earlier comment, but the main concern seemed to be the web.

The web has this feature as of [v1.95.0](https://github.com/immich-app/immich/releases/tag/v1.95.0) (PR #6753) as a user setting 🎉:
![Screenshot_20241201_142449_Brave](https://github.com/user-attachments/assets/6b633e02-532b-42ad-89a1-54f6f196c2e7)

<br>

### Extra notes:

> 1. An option to disable thumbnail creation for commonly supported file formats - similar to how it's already possible to disable transcoding for video.

I think this is not implemented.

<br>

> 2. Generating thumbnails on demand, based on the viewport size and format support of the client

Not implemented. (Probably unlikely due to difficulty and priority.)

<br>

> 3. Supporting a more current format for thumbnails, like AVIF - jpeg lacks support for higher bit depth or HDR

WebP is now a supported format.

## I will close this discussion as it seems the main feature asked for was added and these 3 subpoints were suggested as alternatives.

(Let me know if there is a good reason to reopen.)
